modernize Request and remove jsonRequest

mmatczuk opened the pull request:
jsonRequest is replaced with json tags + omitempty to avoid pointers.

We introduce:
* TunnelDomains,
* DirectDomains,
* DenyDomains,
* TLSPassthroughDomains and deprecate the old and/or unused fields.

This would support both 4.9 and 5.0 until we remove the deprecated fields.

mmatczuk committed:
Remove jsonRequest

jsonRequest is replaced with json tags and omitempty where we had pointers.

mmatczuk commented:
Something is not working in the backend I need to see why.

mmatczuk commented:
My bad boolean values cannot be omitempty. Works now.

waggledans commented on client_types.go:
```diff
@@ -33,51 +33,35 @@ type Metadata struct {
-	// optional features and flags.
-	ExtraInfo string
+	TunnelDomains         []string `json:"tunnel_domains,omitempty"`
+	DirectDomains         []string `json:"direct_domains,omitempty"`
```
The `direct_domains` value is expected to be a list of domains (and not regexps) for SC-4. We have quite a mess rn but we'd clean it up.

OK (messy but we fix that after the client is updated):

`deny_domains` - regexps for sc5
`fast_fail_regexps` - regexps for sc4
`tls_passthrough_domains` - regexps for sc5
`no_ssl_bump_domains` - domains for sc4
`tunnel_domains` - regexps for sc5
`domain_names` - domains for sc4

Not OK:
`direct_domains` - could be domain or regexp.

We would need to ensure the backend / REST API don't validate `direct_domains`  for SC5. It's doable but for the transition, we may need to add a new `direct_regexps` that would contain regexps

**UPDATE**: Alana suggested setting the `experimental` flag to let the backend know that the values are intended for sc-5 releases after the latest 5.0.0-alpha8 (ie regexps). After the server is fully updated and stable, we can remove the `experimental` flag.

mmatczuk replied:
Agreed. We need to do some requests marking / versioning, adding `direct_regexps` just increases the entropy.

The `experimental` flag will not work because it's passed from server to client.

I think we can use the new `--metadata` flag for that, the `--metadata` values will be put to [Metadata.Extra field](https://github.com/saucelabs/tunnelrest-go/blob/2ae2aaeb47b5102d5cdb96a86b4c646ea10a9312/client_types.go#L26) so we would have `--metadata api=v5alpha`

WDYT @amckenzie132

mmatczuk replied:
Or simpler just use the Protocol value?

waggledans replied:
> I think we can use the new --metadata flag for that

sounds good. The Protocol value is not enough during the transition. Some `--metadata` value should do it

mmatczuk committed:
Rename Request to CreateTunnelRequestV4

mmatczuk committed:
Rename CreateTunnel to CreateTunnelV4

mmatczuk committed:
Add CreateTunnelRequestV5 and CreateTunnelV5 function

mmatczuk committed:
golangci: sort disabled linters

mmatczuk committed:
golangci: disable wsl

WSL is too opinionated on whitespaces

client_tunnel.go:14:2: only one cuddle assignment allowed before defer statement (wsl)
        defer cancel()
        ^
client_tunnel.go:15:2: return statements should not be cuddled if block has more than two lines (wsl)
        return c.create(ctx, req)
        ^
client_tunnel.go:24:2: only one cuddle assignment allowed before defer statement (wsl)
        defer cancel()
        ^
client_tunnel.go:25:2: return statements should not be cuddled if block has more than two lines (wsl)
        return c.create(ctx, req)
        ^
client_vpn.go:14:2: only one cuddle assignment allowed before defer statement (wsl)
        defer cancel()
        ^
client_vpn.go:15:2: return statements should not be cuddled if block has more than two lines (wsl)
        return c.create(ctx, req)

mmatczuk committed:
Fix CreateTunnelRequestV4 alignment

In addition to reducing the memory footprint it also logically groups fields and makes CreateTunnelRequestV4 and CreateTunnelRequestV5 easier to compare.

mmatczuk commented:
I have reimplemented this so that there is no ambiguity!

We have now `CreateTunnelRequestV4` and `CreateTunnelRequestV5` that are separate types.
The `CreateTunnelRequestV5` does not have `ExtraInfo` nor `VMVersion`.

I suggest to route between v4 and v5 based on the `Protocol` value, or version/release from Metadata. 
If that's too much of a hack we probably should have a separate endpoint path.

amckenzie132 commented:
We can filter based on `release` from Metadata, but it will need to be updated since I've been seeing `5.0.666-dev` for a while.

`ExtraConfig` should be added to `CreateTunnelRequestV5`, as a replacement for both `ExtraInfo` and `VMVersion`
